metals/src/main/scala/scala/meta/internal/metals/codeactions/ImportMissingSymbol.scala

```diff
@@ -192,17 +217,101 @@ class ImportMissingSymbol(compilers: Compilers, buildTargets: BuildTargets)
+          filterImportActions(deduplicated.toSeq.sorted.map(Seq(_)))
+
+        // Generate the "Import all" action if applicable
+        importMissingSymbols(filteredActions)
       }
   }
-
 }
 
 object ImportMissingSymbol {
-
   def title(name: String, packageName: String): String =
     s"Import '$name' from package '$packageName'"
 
   def allSymbolsTitle: String =
     s"Import all missing symbols that are unambiguous"
 }
+
+/**
+ * The QuickFix implementation of ImportMissingSymbol, which is used
+ * when handling missing imports one at a time.
+ */
+class ImportMissingSymbolQuickFix(
+    compilers: Compilers,
+    buildTargets: BuildTargets,
+) extends ImportMissingSymbol(compilers, buildTargets) {
+
+  override val kind: String = ImportMissingSymbolQuickFix.kind
+  override protected val title: String = ImportMissingSymbol.allSymbolsTitle
+
+  override protected def isCallAllowed(
+      file: AbsolutePath,
+      params: l.CodeActionParams,
+  ): Boolean = {
+    // This is used for per-diagnostic fixes, so it's always allowed as long as there
+    // are diagnostics and the range overlaps with them
+    params
+      .getContext()
+      .getDiagnostics()
+      .asScala
+      .exists { diag =>
+        params.getRange().overlapsWith(diag.getRange())
+      }
+  }
+}
+
+object ImportMissingSymbolQuickFix {
+  final val kind: String = l.CodeActionKind.QuickFix
+}
+
+/**
+ * The "source" implementation of ImportMissingSymbol, which is used
+ * for automatically importing all unambiguous symbols for the entire file.
+ *
+ * This implementation only imports symbols that have exactly one unambiguous import source.
+ * Any symbols with multiple possible imports (like Future from scala.concurrent and
+ * java.util.concurrent) will be skipped and require manual resolution.
+ */
+class SourceAddMissingImports(
+    compilers: Compilers,
+    buildTargets: BuildTargets,
+) extends ImportMissingSymbol(compilers, buildTargets) {
+
+  override val kind: String = SourceAddMissingImports.kind
+  override protected val title: String = SourceAddMissingImports.title
+
+  override protected def isCallAllowed(
+      file: AbsolutePath,
+      params: l.CodeActionParams,
+  ): Boolean = {
+    // Only proceed if explicitly called with source.addMissingImports kind
+    isScalaOrSbt(file) && Option(params.getContext.getOnly)
+      .map(_.asScala.toList.contains(kind))
+      .isDefined
+  }
+
+  /**
+   * Override the filtering method to only keep unambiguous imports.
+   * This ensures we only auto-import symbols with exactly one available import.
+   */
+  override protected def filterImportActions(
+      allActions: Seq[Seq[l.CodeAction]]
+  ): Seq[l.CodeAction] = {
+    // Only keep imports that have exactly one solution (unambiguous)
+    val filteredActions = allActions.flatMap { actions =>
+      if (actions.length == 1) actions else Seq.empty
+    }
+
+    pprint.log("source.addMissingImports filteredActions: " + filteredActions)
```

Comment:
```suggestion
```
pprint is turned of on CI so this will not compile unfortunately

Reply:
Line removed